File: data-structures/arrays/python/dynamicarray.py
class DynamicArray:
    """
    Generic implementation of a dynamic array.
    """

    def __init__(self, capacity=1):
        self._size = 0            # how many items in array
        self._capacity = capacity # number of total slots in array
        self._array = [None for _ in range(self._capacity)] 

    def isEmpty(self):
        return self._size == 0

    def indexOf(self, item):
        for i in range(self._size):
            if item == self._array[i]:
                return i
        return -1

    def add(self, item):
        if self._size >= self._capacity:
            if self._capacity == 0:
                self._capacity = 1
            else:
                self._capacity *= 2

            new_arr = DynamicArray(self._capacity)
            for i in range(self._size):
                new_arr[i] = self._array[i]
            self._array = new_arr

        self._array[self._size] = item
        self._size += 1

    def removeAt(self, index):
        if index < 0 or index >= self._size:
            raise IndexError

        item = self._array[index]

        self._size -= 1
        for i in range(index, self._self):
            self._array[i] = self._array[i+1]

        return item

    def remove(self, item):
        index = self.indexOf(item)
        if index == -1:
            return False
        else:
            self.removeAt(index)
            return True

    def __len__(self):
        return self._size

    def __contains__(self, item):
        return self.indexOf(item) != -1

    def __getitem__(self, index):
        return self._array[index]

    def __setitem__(self, index, item):
        self._array[index] = item

    def __str__(self):
        return "[" + ",".join([str(x) for x in self._array[0:self._size]]) + "]"

File: data-structures/arrays/cpp/DynamicArray.cpp
#include "DynamicArray.h"

DynamicArray::DynamicArray() {
    m_capacity = INITIAL_CAPACITY;
    m_size = 0;
    m_array = new int[m_capacity];
}

DynamicArray::~DynamicArray() {
    delete [] m_array;
}

int DynamicArray::size() const { return m_size; }

bool DynamicArray::isEmpty() const { return size() == 0; }

int DynamicArray::indexOf(const int item) const {
    for (int i = 0; i < m_size; i++) {
        if (m_array[i] == item) return i;
    }
    return -1;
}

bool DynamicArray::contains(const int item) const {
    return indexOf(item) != -1;
}

int DynamicArray::get(const int index) {
    assert(0 <= index && index < m_size);
    return m_array[index];
}

void DynamicArray::add(const int item) {
    // resize if size is at least capacity.
    if (m_size >= m_capacity) {
        m_capacity *= GROWTH_FACTOR;
        int *temp = new int[m_capacity];
        std::copy(m_array, m_array + m_size, temp);
        delete [] m_array;
        m_array = temp;
    }

    m_array[m_size++] = item;
}

int DynamicArray::removeAt(const int index) {
    assert(0 <= index && index < m_size);
    int item = m_array[index];
    m_size--;
    for (int i = index; i < m_size; i++) {
        m_array[i] = m_array[i + 1];
    }
    return item;
}

bool DynamicArray::remove(int item) {
    for (int i = 0; i < m_size; i++) {
        if (m_array[i] == item) {
            this->removeAt(i);
            return true;
        }
    }
    return false;
}

void DynamicArray::pretty_print() const {
    std::cout << "[";

    for (int i = 0; i < m_size; i++) {
        if (i == m_size - 1) {
            std::cout << m_array[i];
        }
        else {
            std::cout << m_array[i] << ", ";
        }
    }

    std::cout << "]\n";
}
